### PR TITLE
Big (hopefully final) ADAR1000 update

### PR DESCRIPTION
**Pull request by wsapia-adi**
- Add missing ADAR1000 attributes
- Add more properties & methods for control of an array as a single entity (attenuators, gains, phases, steering, etc.)
- Nest adar1000_channel class under adar1000 to allow for simpler access to the elements in the array (can be accessed via element number in the array class, or from the parent ADAR1000 object)
- Add support for LTC2314-14
- Add LTC2314-14 and one_bit_adc_dac to the documentation

### CODE REVIEW

**Comment by tfcollins** [comment]
Sorry for the delay. Got the go-ahead from Mircea.

1. Squash the formatting/cleanup commits
2. Doc is generating warnings
```
eading sources... [100%] support                                                                                     
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:21: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:25: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:32: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:19: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:21: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:26: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:30: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/one_bit_adc_dac.py:docstring of adi.one_bit_adc_dac.one_bit_adc_dac:4: WARNING: Unexpected indentation.
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
```
3. For the singleton adar1000 class, can you add a note to each input parameter in regards to which are needed only for the array implementation. It can be a bit overwhelming if you just have a single eval board

File: adi/adar1000.py [hunk 1]
```diff
@@ -41,31 +43,374 @@ class adar1000(attribute, context_manager):
...
+                | Channel 2: Element # 14
+                | Channel 3: Element # 13
+                | Channel 4: Element # 9
     """
 
     _device_name = ""
+    _BIAS_CODE_TO_VOLTAGE_SCALE = -0.018824
+
+    class adar1000_channel:
+        """Class for each channel of the ADAR1000. This class is not meant
```
Comment (tfcollins) [inline adi/adar1000.py:82]: Due to the instantiation comment can you add a check in __init__ like: 
```
if type(self) == adar1000_channel:
    raise something
```

File: adi/adar1000.py [hunk 2]
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
...
+        for element in self.elements.values():
+            # Calculate the row and column phases
+            column_phase = element.column * azimuth_phi
+            row_phase = element.row * elevation_phi
+
+            element.tx_phase = column_phase + row_phase
+
+        # Latch in the new phases
+        self.latch_tx_settings()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1687]: Remove

File: adi/adar1000.py [hunk 3]
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
...
+        for element in self.elements.values():
+            # Calculate the row and column phases
+            column_phase = element.column * azimuth_phi
+            row_phase = element.row * elevation_phi
+
+            element.rx_phase = column_phase + row_phase
+
+        # Latch in the new phases
+        self.latch_rx_settings()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1656]: Remove

File: adi/adar1000.py [hunk 4]
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
...
+        """ Latch in new Gain/Phase settings for the Rx """
+        for device in self.devices.values():
+            device.latch_rx_settings()
+        return
+
+    def latch_tx_settings(self):
+        """ Latch in new Gain/Phase settings for the Tx """
+        for device in self.devices.values():
+            device.latch_tx_settings()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1625]: Remove

File: adi/adar1000.py [hunk 5]
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
...
+        """ Initialize the ADAR1000s in the array """
+        for device in self.devices.values():
+            device.initialize()
+        return
+
+    def latch_rx_settings(self):
+        """ Latch in new Gain/Phase settings for the Rx """
+        for device in self.devices.values():
+            device.latch_rx_settings()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1619]: Remove

File: adi/adar1000.py [hunk 6]
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
...
+
+        return az_phi, el_phi
+
+    """ Public Methods """
+
+    def initialize_devices(self):
+        """ Initialize the ADAR1000s in the array """
+        for device in self.devices.values():
+            device.initialize()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1613]: Remove

File: adi/adar1000.py (hunk 1 above)
Comment (wsapia-adi) [reply adi/adar1000.py:82]: I can add this, but what do you think about just hiding it with a leading "_"?

File: adi/adar1000.py (hunk 2 above)
Comment (wsapia-adi) [reply adi/adar1000.py:1687]: Why remove all of these here and below? If these aren't there, none of the new values will take effect...

**Comment by wsapia-adi** [comment]
1. Tried to squash all the formatting and miscellaneous commits, let me know if this is clean enough
2. Not sure what's causing this?
3. Added more info to the docstrings, clearer?
4. Added example files for single and array instances

File: adi/adar1000.py (hunk 1 above)
Comment (tfcollins) [reply adi/adar1000.py:95]: Added a _ prefix is fine.

File: adi/adar1000.py (hunk 2 above)
Comment (tfcollins) [reply adi/adar1000.py:1689]: What is the purpose of the return statement?

Comment (wsapia-adi) [reply adi/adar1000.py:1689]: I didn't realize you were referring to the return, I thought you were saying to remove the latch command. I get that it doesn't return anything, but it works the same either way. Just a habit. I can remove it.

File: adi/adar1000.py (hunk 1 above)
Comment (wsapia-adi) [reply adi/adar1000.py:95]: I should undo that then. It needs to show up in the doc so that people can see what properties/methods are available.

Comment (wsapia-adi) [reply adi/adar1000.py:95]: I'm not sure I can add your check that you originally suggested, as I believe that will always fail? During __init__, won't self always equal adar1000_channel?

File: adi/adar1000.py [hunk 7]
```diff
@@ -85,693 +440,820 @@ def __init__(self, uri="", chip_id="csb1_chip1", context_handle=None):
...
+            channel.rx_gain = 0x7F
+            channel.rx_phase = 0
+            channel.tx_attenuator = False
+            channel.tx_gain = 0x7F
+            channel.tx_phase = 0
+
+        # Latch in the new settings
+        self.latch_rx_settings()
+        self.latch_tx_settings()
+        return
```
Comment (tfcollins) [inline adi/adar1000.py:1152]: Remove. I think that's the last one and we should be good to go.

**Pull request merged**